File: src/main/java/com/weeth/domain/board/domain/entity/Board.java
package com.weeth.domain.board.domain.entity;

import jakarta.persistence.Column;
import jakarta.persistence.EntityListeners;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import java.util.List;
import com.weeth.domain.board.application.dto.NoticeDTO;
import com.weeth.domain.board.application.dto.PostDTO;
import com.weeth.domain.comment.domain.entity.Comment;
import com.weeth.domain.user.domain.entity.User;
import com.weeth.global.common.entity.BaseEntity;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.experimental.SuperBuilder;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.jpa.domain.support.AuditingEntityListener;

@Getter
@MappedSuperclass
@EntityListeners(AuditingEntityListener.class)
@SuperBuilder
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Slf4j
public class Board extends BaseEntity {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    public Long id;

    private String title;

    @Column(columnDefinition = "TEXT")
    private String content;

    @ManyToOne
    @JoinColumn(name = "user_id")
    private User user;

    private Integer commentCount;

    @PrePersist
    public void prePersist() {
        commentCount = 0;
    }

    public void decreaseCommentCount() {
        if (commentCount > 0) {
            commentCount--;
        }
    }

    public void increaseCommentCount() {
        commentCount++;
    }

    public void updateCommentCount(List<Comment> comments) {
        this.commentCount = (int) comments.stream()
                .filter(comment -> !comment.getIsDeleted())
                .count();
    }

    public void updateUpperClass(NoticeDTO.Update dto) {
        this.title = dto.title();
        this.content = dto.content();
    }

    public void updateUpperClass(PostDTO.Update dto) {
        if (dto.title() != null) this.title = dto.title();
        if (dto.content() != null) this.content = dto.content();
    }

    public void updateUpperClass(PostDTO.UpdateEducation dto) {
        if (dto.title() != null) this.title = dto.title();
        if (dto.content() != null) this.content = dto.content();
    }
}

File: src/main/kotlin/com/weeth/domain/board/domain/entity/Board.kt
package com.weeth.domain.board.domain.entity

import com.weeth.domain.board.domain.converter.BoardConfigConverter
import com.weeth.domain.board.domain.enums.BoardType
import com.weeth.domain.board.domain.vo.BoardConfig
import com.weeth.domain.club.domain.entity.Club
import com.weeth.domain.club.domain.enums.MemberRole
import com.weeth.global.common.entity.BaseEntity
import jakarta.persistence.Column
import jakarta.persistence.Convert
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table

@Entity
@Table(name = "board")
class Board(
    club: Club,
    name: String,
    description: String,
    type: BoardType,
    config: BoardConfig = BoardConfig(),
) : BaseEntity() {
    init {
        require(name.isNotBlank()) { "게시판 이름은 공백이 될 수 없습니다" }
        require(description.isNotBlank()) { "게시판 설명은 공백이 될 수 없습니다" }
        require(type != BoardType.ALL) { "ALL은 가상 타입으로 게시판을 생성할 수 없습니다" }
    }

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "club_id", nullable = false)
    var club: Club = club
        private set

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long = 0L
        private set

    @Column(nullable = false)
    var name: String = name
        private set

    @Column(nullable = false, length = 500)
    var description: String = description
        private set

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var type: BoardType = type
        private set

    @Column(columnDefinition = "JSON") // Json 속성 사용으로 인한 커스텀 컨버터 적용
    @Convert(converter = BoardConfigConverter::class)
    var config: BoardConfig = config
        private set

    @Column(nullable = false)
    var displayOrder: Int = 0
        private set

    @Column(nullable = false)
    var isDeleted: Boolean = false
        private set

    val isCommentEnabled: Boolean
        get() = config.commentEnabled

    val isAdminOnly: Boolean
        get() = config.writePermission.isAdminOrLead()

    fun isAccessibleBy(memberRole: MemberRole): Boolean = memberRole.isAdminOrLead() || !config.isPrivate

    fun canWriteBy(memberRole: MemberRole): Boolean =
        isAccessibleBy(memberRole) && (memberRole.isAdminOrLead() || !isAdminOnly)

    fun updateConfig(newConfig: BoardConfig) {
        config = newConfig
    }

    fun rename(newName: String) {
        require(newName.isNotBlank()) { "게시판 이름은 공백이 될 수 없습니다." }
        name = newName
    }

    fun updateDescription(newDescription: String) {
        require(newDescription.isNotBlank()) { "게시판 설명은 공백이 될 수 없습니다." }
        description = newDescription
    }

    fun markDeleted() {
        isDeleted = true
    }

    fun restore() {
        isDeleted = false
    }

    fun reorder(newOrder: Int) {
        require(newOrder >= 0) { "순서는 0 이상이어야 합니다." }
        displayOrder = newOrder
    }
}
